pipenv/core.py

```diff
@@ -395,18 +431,20 @@ def abort():
             err=True,
         )
         # check for python installers
-        from .vendor.pythonfinder.environment import PYENV_INSTALLED, ASDF_INSTALLED
         from .installers import Pyenv, Asdf, InstallerError
 
         # prefer pyenv if both pyenv and asdf are installed as it's
         # dedicated to python installs so probably the preferred
         # method of the user for new python installs.
-        if PYENV_INSTALLED and not PIPENV_DONT_USE_PYENV:
-            installer = Pyenv("pyenv")
-        elif ASDF_INSTALLED and not PIPENV_DONT_USE_ASDF:
-            installer = Asdf("asdf")
-        else:
-            installer = None
+        installer = None
+        if not PIPENV_DONT_USE_PYENV:
```

Comment:
It's not clear to me how this is an improvement / what is changing about the logic here, can you explain what problem this solves and provide an example of how to cause it on the master branch?

Reply:
Assuming neither of the `DONT_USE` variables are set:


**On master:**
If `PYENV_ROOT` is set, but `pyenv` the binary is not on `PATH`, the stacktrace posted in #4245 occurs.

**In this PR:**
If `PYENV_ROOT` is set, but `pyenv` the binary is not on `PATH`, the following logic is used:
- If a `pyenv` executable can be located in one of three locations, it will be used.
- If not, Pipenv will search for an `asdf` binary using the equivalent search process.

This provides two benefits:
1. `pyenv`/`asdf` is searched for more aggressively and don't need to be on `PATH`.
2. Fallback from `pyenv` to `asdf` will occur more reliably.

Reply:
Thanks for the cogent explanation, this clarifies the issue here and what the change accomplishes. I am ok with this -- thanks for putting it together and for explaining!